bokehjs/src/lib/core/util/templating.ts

```diff
@@ -25,7 +25,8 @@ function _format_number(num: string | number): string {
     return `${num}`  // get strings for categorical types
 }
 
-export function replace_placeholders(str: string, data_source: ColumnarDataSource, i: number,
+
+export function replace_placeholders(str: string, data_source: ColumnarDataSource, i : any,
```

Comment:
Going to make a stab at suggesting appropriate typings (but I'm definitely not an expert).

``i: number | [number, number, number, number]`` should be sufficient

Reply:
@mattpap I tried Philipp suggestion above (which looks right to me) without any luck. I get:

```
TS2538: Type 'null' cannot be used as an index type.
```

I've tried a number of variants, including some that allow ``null`` but none of them work. This column data source index is now the only new thing introduced in this PR with the ``any`` type.

Reply:
Defining ``ImageIndex`` helped solve this problem for ``_render_tooltips`` but I am still having trouble using it in ``replace_placeholders``....